### PR TITLE
Resubmit any operations not yet in "Pending" state when an idempotency key clash is detected

### PR DESCRIPTION
**Pull request by matthew1001**
Fixes https://github.com/hyperledger/firefly/issues/1219 - see that issue and associated Miro board for complete detail on the changes proposed under this PR.

### CODE REVIEW

**Review by peterbroadhurst — COMMENTED** [review]
Main feedback at this stage @matthew1001, is that it seems like we need a utility in the operation manager to do this, that has the boilerplate code in it.

I wonder even if it could be embedded in the `SubmitNewTransaction` helper?

Then that could just return that the TX was re-submitted with the previous values.... and all the code could just check that return code, and return early rather than continuing to build new operations.

**Comment by matthew1001** [comment]
Seeing a test coverage drop in `nextpin_sql.go` which I'm not sure anything in this PR has affected, but looking into it

**Review by peterbroadhurst — APPROVED** [review]
This looks great to me @matthew1001 👍 
Thanks for taking the time to minimize the amount of code in each of the idempotency checks.

**Comment by peterbroadhurst** [comment]
I do see this error in the E2E, which suggests a panic:
```
        	Error:      	Received unexpected error:
        	            	unexpected EOF
        	            	```

**Review by peterbroadhurst — CHANGES_REQUESTED** [review]
There's one E2E test failure that shows a `404` where the test expects a `409` that needs digging into:
```
    restclient.go:119: Headers: map[Content-Length:[37] Content-Type:[application/json] Date:[Tue, 28 Mar 2023 19:31:16 GMT] Vary:[Origin]]
    restclient.go:811: 
        	Error Trace:	/home/runner/work/firefly/firefly/test/e2e/runners/restclient.go:811
        	            				/home/runner/work/firefly/firefly/test/e2e/runners/ethereum_contracts.go:231
        	Error:      	Not equal: 
        	            	expected: 409
        	            	actual  : 404
        	Test:       	TestEthereumMultipartyE2ESuite/TestFFIInvokeMethod
        	Messages:   	POST /namespaces/default/contracts/invoke [404]: {"error":"FF00164: No result found"}
```

Here's the relevant section of logs:
```
[34mfirefly_core_0_1  |[0m [2023-03-28T19:31:16.498Z] ERROR SQL insert failed: UNIQUE constraint failed: transactions.namespace, transactions.idempotency_key sql=[ INSERT INTO transactions (id,ttype,namespace,created,idempotency_key,blockchain_ids) VALUES ($1,$2,$3,$4,$5,$6) ]: UNIQUE constraint failed: transactions.namespace, transactions.idempotency_key dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
[34mfirefly_core_0_1  |[0m [2023-03-28T19:31:16.498Z] DEBUG SQL-> query transactions dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
[34mfirefly_core_0_1  |[0m [2023-03-28T19:31:16.498Z] DEBUG SQL<- query transactions dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
[34mfirefly_core_0_1  |[0m [2023-03-28T19:31:16.498Z] DEBUG SQL-> query operations dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
[34mfirefly_core_0_1  |[0m [2023-03-28T19:31:16.498Z] DEBUG SQL<- query operations dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
[34mfirefly_core_0_1  |[0m [2023-03-28T19:31:16.498Z]  WARN SQL! transaction rollback dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
[34mfirefly_core_0_1  |[0m [2023-03-28T19:31:16.499Z]  INFO <-- POST /api/v1/namespaces/default/contracts/invoke [404] (0.86ms): FF00164: No result found httpreq=d361Qotq pid=1 req=7rVqxwVL
```

**Comment by matthew1001** [comment]
> There's one E2E test failure that shows a `404` where the test expects a `409` that needs digging into:
> 
> ```
>     restclient.go:119: Headers: map[Content-Length:[37] Content-Type:[application/json] Date:[Tue, 28 Mar 2023 19:31:16 GMT] Vary:[Origin]]
>     restclient.go:811: 
>         	Error Trace:	/home/runner/work/firefly/firefly/test/e2e/runners/restclient.go:811
>         	            				/home/runner/work/firefly/firefly/test/e2e/runners/ethereum_contracts.go:231
>         	Error:      	Not equal: 
>         	            	expected: 409
>         	            	actual  : 404
>         	Test:       	TestEthereumMultipartyE2ESuite/TestFFIInvokeMethod
>         	Messages:   	POST /namespaces/default/contracts/invoke [404]: {"error":"FF00164: No result found"}
> ```
> 
> Here's the relevant section of logs:
> 
> ```
> �[34mfirefly_core_0_1  |�[0m [2023-03-28T19:31:16.498Z] ERROR SQL insert failed: UNIQUE constraint failed: transactions.namespace, transactions.idempotency_key sql=[ INSERT INTO transactions (id,ttype,namespace,created,idempotency_key,blockchain_ids) VALUES ($1,$2,$3,$4,$5,$6) ]: UNIQUE constraint failed: transactions.namespace, transactions.idempotency_key dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
> �[34mfirefly_core_0_1  |�[0m [2023-03-28T19:31:16.498Z] DEBUG SQL-> query transactions dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
> �[34mfirefly_core_0_1  |�[0m [2023-03-28T19:31:16.498Z] DEBUG SQL<- query transactions dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
> �[34mfirefly_core_0_1  |�[0m [2023-03-28T19:31:16.498Z] DEBUG SQL-> query operations dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
> �[34mfirefly_core_0_1  |�[0m [2023-03-28T19:31:16.498Z] DEBUG SQL<- query operations dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
> �[34mfirefly_core_0_1  |�[0m [2023-03-28T19:31:16.498Z]  WARN SQL! transaction rollback dbtx=naHfZ4Wu httpreq=d361Qotq pid=1 req=7rVqxwVL
> �[34mfirefly_core_0_1  |�[0m [2023-03-28T19:31:16.499Z]  INFO <-- POST /api/v1/namespaces/default/contracts/invoke [404] (0.86ms): FF00164: No result found httpreq=d361Qotq pid=1 req=7rVqxwVL
> ```

Think e2e is now fixed @peterbroadhurst

**Comment by matthew1001** [comment]
> Hmm, I'm seeing build failures with

OK, all good. Think git checkout was playing up. Build & e2e tests now passing green @peterbroadhurst

**Review by peterbroadhurst — APPROVED** [review]

File: internal/operations/manager.go [hunk 1]
```diff
@@ -134,11 +159,19 @@ func (om *operationsManager) RunOperation(ctx context.Context, op *core.Prepared
 			ErrorMessage:   err.Error(),
 			Output:         outputs,
 		})
-	} else if complete {
+	} else {
+		// No error so move us from "Initialized" to "Pending"
+		newState := core.OpStatusPending
```
Comment (awrichar) [inline internal/operations/manager.go:164]: This is introducing a race condition - see https://github.com/hyperledger/firefly/issues/1256

**Pull request merged**